test/production/graceful-shutdown/index.test.ts

```diff
@@ -0,0 +1,236 @@
+      expect(isAppRunning(app)).toBe(true)
+
+      // App finally shuts down
+      await appKilledPromise
+      expect(isAppRunning(app)).toBe(false)
+    })
+  } else {
+    it('should wait for requests to complete before exiting', async () => {
+      const appKilledPromise = once(app, 'exit')
+
+      let responseResolved = false
+      const resPromise = fetchViaHTTP(appPort, '/api/long-running')
+        .then((res) => {
+          responseResolved = true
+          return res
+        })
+        .catch(() => {})
+
+      // yield event loop to kick off request before killing the app
+      await waitFor(20)
+      process.kill(app.pid, 'SIGTERM')
+      expect(isAppRunning(app)).toBe(true)
+
+      // Long running response should still be running after a bit
+      await waitFor(LONG_RUNNING_MS / 2)
+      expect(isAppRunning(app)).toBe(true)
+      expect(responseResolved).toBe(false)
+
+      // App responds as expected without being interrupted
+      const res = await resPromise
+      assertDefined(res)
+      expect(res.status).toBe(200)
+      expect(await res.json()).toStrictEqual({ hello: 'world' })
+
+      // App is still running briefly after response returns
+      expect(isAppRunning(app)).toBe(true)
+      expect(responseResolved).toBe(true)
+
+      // App finally shuts down
+      await appKilledPromise
+      expect(isAppRunning(app)).toBe(false)
+    })
+
+    describe('should not accept new requests during shutdown cleanup', () => {
+      it('when request is made before shutdown', async () => {
+        const appKilledPromise = once(app, 'exit')
+
+        const resPromise = fetchViaHTTP(appPort, '/api/long-running')
+
+        // yield event loop to kick off request before killing the app
+        await waitFor(20)
+        process.kill(app.pid, 'SIGTERM')
+        expect(isAppRunning(app)).toBe(true)
+
+        // Long running response should still be running after a bit
+        await waitFor(LONG_RUNNING_MS / 2)
+        expect(isAppRunning(app)).toBe(true)
+
+        // Second request should be rejected
+        await expect(
+          fetchViaHTTP(appPort, '/api/long-running')
+        ).rejects.toThrow()
+
+        // Original request responds as expected without being interrupted
+        await expect(resPromise).resolves.toBeDefined()
+        const res = await resPromise
+        expect(res.status).toBe(200)
+        expect(await res.json()).toStrictEqual({ hello: 'world' })
+
+        // App is still running briefly after response returns
+        expect(isAppRunning(app)).toBe(true)
+
+        // App finally shuts down
+        await appKilledPromise
+        expect(isAppRunning(app)).toBe(false)
+      })
+
+      it('when there is no activity', async () => {
+        const appKilledPromise = once(app, 'exit')
+
+        process.kill(app.pid, 'SIGTERM')
+        expect(isAppRunning(app)).toBe(true)
+
+        // yield event loop to allow server to start the shutdown process
+        await waitFor(20)
```

Comment:
I was running into some issues with these tests. Waiting 20 ms seems to be the sweet spot but I'm a little worried there could be race conditions causing this to sometimes fail.